Implement ser_json_temporal config option

ollz272 opened the pull request:
<!-- Thank you for your contribution! -->

Implements the first part of https://github.com/pydantic/pydantic/pull/11504, which is the ser_json_temporal config option.

We now have a `ser_json_temporal` part to the config, which accepts ['iso8601', 'seconds', 'milliseconds']. These have been implemented as designed in the original pydantic pull request ('iso8601' being obvious, seconds and milliseconds each returning the number of seconds and milliseconds in a float).

This has been done by implementing the serializers for each of the specified types (datetime, date, time, timedelta).

Special case for 'timedelta' where if the old config option is set, it will be used instead of ser_json_temporal.

<!-- Please give a short summary of the changes. -->

## Related issue number

 https://github.com/pydantic/pydantic/pull/11504

## Checklist

* [x ] Unit tests for the changes exist
* [x ] Documentation reflects the changes where applicable
* [ ] Pydantic tests pass with this `pydantic-core` (except for expected changes)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**


Selected Reviewer: @sydney-runkle

ollz272 commented:
please review

ollz272 commented on src/input/datetime.rs:
```diff
@@ -131,6 +131,93 @@ impl EitherTimedelta<'_> {
         }
     }
+
+    pub fn total_seconds(&self) -> PyResult<f64> {
```
This has been copied from the original time we tried to implement float seconds/milliseconds into the lib 😄

ollz272 commented on src/serializers/config.rs:
```diff
@@ -18,29 +23,52 @@ use super::errors::py_err_se_err;
 pub(crate) struct SerializationConfig {
     pub timedelta_mode: TimedeltaMode,
+    pub temporal_mode: TemporalMode,
+    prefer_timedelta_mode: bool,
```
Im not super happy with this code being repeated here & in timedelta.rs... but im not sure if theres a better way of doing things.

davidhewitt replied:
Here I would suggest having `timedelta_mode` be `Option<TimedeltaMode>`, which should make it obvious what's going on.

ollz272 commented on src/serializers/config.rs:
```diff
@@ -91,6 +119,14 @@ serialization_mode! {
     Float => "float",
 }
 
+serialization_mode! {
```
The new ser options for temporal types

ollz272 commented on src/serializers/config.rs:
```diff
@@ -163,6 +199,174 @@ impl TimedeltaMode {
 }
 
+impl TemporalMode {
+    pub fn datetime_to_json(self, py: Python, datetime: &Bound<'_, PyDateTime>) -> PyResult<PyObject> {
```
Implemented all the key bits here. Felt like the right place, happy to move

ollz272 commented on src/serializers/type_serializers/datetime_etc.rs:
```diff
@@ -2,21 +2,35 @@ use std::borrow::Cow;
+    pydate_as_date(py_date).map(|dt| dt.timestamp() as f64)
+}
+pub(crate) fn date_to_milliseconds(py_date: &Bound<'_, PyDate>) -> PyResult<f64> {
+    pydate_as_date(py_date).map(|dt| dt.timestamp_ms() as f64)
```
The speedate functions return an i64, so casting to f64 so we return floats like the task says can change if needed.

ollz272 commented on src/serializers/type_serializers/datetime_etc.rs:
```diff
@@ -37,93 +59,215 @@ fn downcast_date_reject_datetime<'a, 'py>(py_date: &'a Bound<'py, PyAny>) -> PyR
     Err(PydanticSerializationUnexpectedValue::new_from_msg(None).to_py_err())
 }
 
-macro_rules! build_serializer {
```
One thing is that this meant this marco wasn't useful anymore. But now all of these serializers follow a very similar pattern. Im not super clued in with marcos so maybe these can all be rewritten to use a marco, but leaving as is for now.

davidhewitt replied:
I would prefer use a generic to a macro tbh, can help think of ways to factor that out if needed?

ollz272 replied:
@davidhewitt done 😄

ollz272 commented on src/serializers/type_serializers/timedelta.rs:
```diff
@@ -26,7 +33,28 @@ impl BuildSerializer for TimeDeltaSerializer {
-        Ok(Self { timedelta_mode }.into())
+        let temporal_mode = TemporalMode::from_config(config)?;
+
+        let prefer_timedelta_mode = config
```
Again, not sure on my implementation here, its all i could really think of. Happy to take feedback

ollz272 commented on tests/serializers/test_any.py:
```diff
@@ -722,3 +722,385 @@ class MyEnum(Enum):
     assert v.to_json(1) == b'1'
+
+
+@pytest.mark.parametrize(
```
Can add more tests, hopefully the abundance of tests here gives confidence this is all working as intended 😅

davidhewitt commented on src/serializers/type_serializers/datetime_etc.rs:
```diff
@@ -2,21 +2,35 @@ use std::borrow::Cow;
     pydatetime_as_datetime(py_dt).map(|dt| dt.to_string())
 }
 
+pub(crate) fn datetime_to_seconds(py_dt: &Bound<'_, PyDateTime>) -> PyResult<f64> {
+    pydatetime_as_datetime(py_dt).map(|dt| dt.timestamp() as f64)
+}
+
+pub(crate) fn datetime_to_milliseconds(py_dt: &Bound<'_, PyDateTime>) -> PyResult<f64> {
+    pydatetime_as_datetime(py_dt).map(|dt| dt.timestamp_ms() as f64)
+}
```
These two seem wrong to me, probably need to go through the functions on `EitherTimedelta`?

davidhewitt commented on src/serializers/type_serializers/datetime_etc.rs:
```diff
@@ -25,6 +39,14 @@ pub(crate) fn time_to_string(py_time: &Bound<'_, PyTime>) -> PyResult<String> {
     pytime_as_time(py_time, None).map(|dt| dt.to_string())
 }
 
+pub(crate) fn time_to_seconds(py_time: &Bound<'_, PyTime>) -> PyResult<f32> {
+    pytime_as_time(py_time, None).map(|t| t.total_seconds() as f32)
+}
+
+pub(crate) fn time_to_milliseconds(py_time: &Bound<'_, PyTime>) -> PyResult<f32> {
+    pytime_as_time(py_time, None).map(|t| t.total_ms() as f32)
+}
```
I feel like again here given we return floats we should be using similar functions to the `EitherTimedelta` stuff, to add the microseconds on as a fraction.

davidhewitt commented on src/serializers/config.rs:
```diff
@@ -163,6 +199,174 @@ impl TimedeltaMode {
+
+    pub fn date_to_json(self, py: Python, date: &Bound<'_, PyDate>) -> PyResult<PyObject> {
+        match self {
+            Self::Iso8601 => Ok(date_to_string(date)?.into_py_any(py)?),
+            Self::Seconds => Ok(date_to_seconds(date)?.into_py_any(py)?),
+            Self::Milliseconds => Ok(date_to_milliseconds(date)?.into_py_any(py)?),
```
You can drop outer `Ok` from a lot of the functions here, e.g.:

```suggestion
            Self::Iso8601 => date_to_string(date)?.into_py_any(py),
            Self::Seconds => date_to_seconds(date)?.into_py_any(py),
            Self::Milliseconds => date_to_milliseconds(date)?.into_py_any(py),
```

davidhewitt commented on tests/serializers/test_any.py:
```diff
@@ -722,3 +722,385 @@ class MyEnum(Enum):
+            'iso8601',
+        ),
+        (
+            time(3, 14, 1, 59263),
+            11641.0,
+            b'11641.0',
+            {'11641': 'foo'},
+            b'{"11641":"foo"}',
+            'seconds',
+        ),
+        (
+            time(3, 14, 1, 59263),
+            11641059.0,
+            b'11641059.0',
+            {'11641059': 'foo'},
+            b'{"11641059":"foo"}',
```
e.g. here I think we want to get the microsecond precision correct.

ollz272 commented:
@davidhewitt I think i've addressed all your comments bar the temporal_mode bits on the timedelta, keep trying things but nothing seems right so i think im doing something wrong!

ollz272 commented:
@davidhewitt actually, think i got the gist of it, didn't use the option<TimedeltaMode> in the end just create a method that turns the timedeltamode into a temporal mode for timedelta ser. Think this is everything resolved now

ollz272 commented:
@davidhewitt was there anything else you wanted me to have a look at here?

davidhewitt commented:
Sorry for dropping the ball here - will seek to complete review of both these PRs by EOD tomorrow.
